src/Akka.Streams.Kafka/Stages/Producers/DefaultProducerStage.cs

```diff
@@ -171,39 +193,60 @@ public override void PostStop()
         
         protected virtual void PostSend(IEnvelope<K, V, P> msg) { }
 
-        private Action<DeliveryReport<K, V>> BuildSendCallback<TResult>(TaskCompletionSource<TResult> completion, Action<DeliveryReport<K, V>> onSuccess)
+        private void OnProduceFailure(Exception ex)
+        {
+            switch (_decider(ex))
+            {
+                case Directive.Stop:
+                    CloseAndFailStage(ex);
+                    Log.Error(ex, $"Producer.Produce threw an exception: {ex.Message}");
+                    break;
+                default:
+                {
+                    Log.Debug($"This exception has been handled by the Supervision Decider: {ex.Message}");
+                    if (ex is ProduceException<K, V> pEx)
+                    {
+                        var error = pEx.Error;
+                        if (error.IsFatal)
+                        {
+                            // if it is a fatal exception, producer needs to be restarted
+                            Producer = _stage.ProducerProvider(null);
+                            Log.Debug($"Producer restarted: {Producer.Name}");
+                        }
+                    }
+                    TryPull(_stage.In);
+                    ConfirmAndCheckForCompletion();
+                    break;
+                }
+            }
+        }
+        
+        private Action<DeliveryReport<K, V>> BuildSendCallback<TResult>(
```

Comment:
Aggregated all callbacks into a single call. Before, each callback methods are decorated with the GetAsyncCallback inside this method. Now all of them are regular methods, we call the GetAsyncCallback decorator on this method instead. Less headache to maintain.